pkg/controlplane/controller/leaderelection/run_with_leaderelection.go

```diff
@@ -50,7 +47,7 @@ func RunWithLeaderElection(ctx context.Context, config *rest.Config, newRunnerFn
 			run(ctx, 1)
 		},
 		OnStoppedLeading: func() {
-			cancel()
+			klog.Info("Stopped leading")
```

Comment:
is OnStartedLeading supposed to be able to be called multiple times? if it is, is it expected to be paired 1:1 with calls to OnStoppedLeading?

Reply:
Checking the code it seems it is only called once https://github.com/kubernetes/kubernetes/pull/127737/files/fff04976265b0186000191a5be76c940e5bce3cf#r1781331967

Since this function does not return errors, I think we need logs to be able to understand the state for troubleshooting

Reply:
Leader transitions should be rare (generally only on failure conditions or restarts/upgrades) so I'm in favor of keeping the log for troubleshooting.

You're right that OnStartedLeading and OnStoppedLeading are called once max, though it should be possible for an apiserver to acquire leader, release it, and acquire the leader again. The code currently doesn't allow reacquiring after release and is a bug which I will submit a fix for https://github.com/kubernetes/kubernetes/issues/127783 but that's orthogonal to the panic issue.

Reply:
> The code currently doesn't allow reacquiring after release and is a bug which I will submit a fix for #127783 but that's orthogonal to the panic issue.

It's not really orthogonal ... OnStoppedLeading is using a variable set by OnStartedLeading... if they can be called concurrently or repeatedly, the current variable sharing is unsafe (which can cause a panic, a data race, or a hung context)

Reply:
Callbacks can not be called concurrently checking the code ... and now it seems you need to `Run()` again once the leadership is lost

Reply:
> Callbacks can not be called concurrently checking the code

I'm... not sure that's true... we call OnStartedLeading in a goroutine so it *could* be concurrent with OnStoppedLeading() (but I agree it's unlikely to be):

```
go le.config.Callbacks.OnStartedLeading(ctx)
```

a simple nil check on `ctx` with a TODO to resolve concurrency issues and multi-call issues in a follow-up should resolve the test flake and not make things worse, and be a smaller change to reason about